### PR TITLE
[PLATFORM-1361] Route namespaces

### PR DESCRIPTION
**Pull request by mondoreale**
In this PR I make it possible to group routes by what I call namespaces. Its sole purpose is to avoid routes like `userpagesFancyRouteToProduct`. Instead you now can do that:

```json
{
    "userpages": {
        "fancyRoute": {
            "product": "/core/products/:id",
            "stream": "/core/streams/:id"
        }
    }
}
```

And use it later like this

```js
import routes from '$routes'

routes.userpages.fancyRoute.product({ id: 13 })
// /core/products/13
```

I also went ahead and applied this to community, contact, and a couple of other routes.

Hope you guys like it!

### CODE REVIEW

File: app/src/links.js [hunk 1]
```diff
@@ -53,21 +53,15 @@ export default {
...
-        medium: routes.communityMedium(),
-        github: routes.communityGithub(),
-        devForum: 'https://forum.streamr.dev/',
-    },
-    contact: {
-        general: 'mailto:contact@streamr.com',
-        media: 'mailto:media@streamr.com',
-        jobs: 'mailto:jobs@streamr.com',
-        labs: 'mailto:labs@streamr.com',
+        trello: routes.community.trello(),
```
Comment (mondoreale) [inline app/src/links.js:56]: Will get rid of those eventually and use `routes` directly.

File: app/src/routes/index.js [hunk 2]
```diff
@@ -26,8 +26,9 @@ type Variables = {
  */
 export const define = (pathstr: string, getVariables: () => Variables) => (params: ?Object, hash?: ?string): string => {
     const route = Object.entries(getVariables()).reduce((acc, [name, value]) => {
-        const val: any = value
-        return pathstr.replace(new RegExp(`<${name}>`, 'g'), val)
+        const val: any = value || ''
+        const strippedValue: string = val.length > 1 ? val.replace(/\/$/, '') : val
+        return acc.replace(new RegExp(`<${name}>`, 'g'), strippedValue)
```
Comment (mondoreale) [inline app/src/routes/index.js:31]: Takes care of redundant slashes and fixes applying variables.

File: app/src/routes/index.js [hunk 3]
```diff
@@ -55,7 +56,11 @@ export const buildRoutes = (paths: Paths, getVariables: () => Variables): Routes
 
         return {
             ...acc,
-            [name]: define(value, getVariables),
+            [name]: typeof value === 'string' ? (
+                define(value, getVariables)
+            ) : (
+                buildRoutes(value, getVariables)
```
Comment (mondoreale) [inline app/src/routes/index.js:62]: Recursion ftw!

**Comment by mondoreale** [comment]
Created [`PLATFORM-1362`](https://streamr.atlassian.net/browse/PLATFORM-1362), FYI.

File: app/src/routes/routes.test.js [hunk 4]
```diff
@@ -14,15 +32,64 @@ describe('route utils', () => {
...
+        it('generates a namespaced route using variabled (slash-squish)', () => {
+            expect(routes.namespace.slashed()).toEqual('https://streamr.network/ns/api/resource/:id')
+            expect(routes.namespace.slashed({
+                id: 13,
+            })).toEqual('https://streamr.network/ns/api/resource/13')
+        })
     })
 
-    describe('defile', () => {
+    describe('define', () => {
```
Comment (fonty1) [inline app/src/routes/routes.test.js:92]: 😄

**Pull request merged**